implementations/utils/gpu.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities to create, delete, and SSH with GPUs."""
+
+import io
+from typing import Iterable, Tuple
+import uuid
+
+from absl import logging
+import airflow
+from airflow.decorators import task, task_group
+from airflow.utils.task_group import TaskGroup
+from apis import gcp_config, test_config
+import fabric
+import google.api_core.exceptions
+import google.auth
+import google.longrunning.operations_pb2 as operations
+from implementations.utils import ssh
+from implementations.utils.vm_api import gpu_api
+import paramiko
+
+
+@task
+def generate_gpu_name(base_gpu_name: str) -> str:
+  # note: GPU vm name need to match regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)', while TPU vm allows '_'.
+  # return f'{base_gpu_name}-{str(uuid.uuid4())}'.replace('_', '-')
+  # If we use the above base_gpu_name in the return, some potion of the can be longer than 61 as in the regex.
+  return f'gpu-{str(uuid.uuid4())}'
+
+
+def create_resource(
+    gpu_name: airflow.XComArg,
+    image_project: str,
+    image_family: str,
+    accelerator,
+    vm_duration,
+    gcp: gcp_config.GCPConfig,
+    ssh_keys: airflow.XComArg,
+) -> Tuple[TaskGroup, airflow.XComArg]:
+  """Request a resource and wait until the nodes are created.
+
+  Args:
```

Comment:
Missing some Args in the comment. Also, wondering the reason that `create_resource_request` instead of `create_resource` method? We don't need to wrap it in a task group if there is only 1 task.

Reply:
Removed the wrap.